include/bredis/impl/async_op.ipp

```diff
@@ -59,46 +59,51 @@ class async_read_op {
 
-    if (error_code) {
-        return callback_(error_code, std::move(result));
-    }
-    auto const_buff = rx_buff_.data();
-    auto begin = Iterator::begin(const_buff);
-    auto end = Iterator::end(const_buff);
-
-    markers::array_holder_t<Iterator> results;
-    results.elements.reserve(replies_count_);
-    size_t cumulative_consumption = 0;
-
-    do {
-        auto from = begin + cumulative_consumption;
-        auto parse_result = Protocol::parse(from, end);
-        auto *parse_error = boost::get<protocol_error_t>(&parse_result);
-        if (parse_error) {
-            auto parse_error_code =
-                Error::make_error_code(bredis_errors::protocol_error);
-            return callback_(parse_error_code, std::move(result));
+    if (!error_code) {
+        auto const_buff = rx_buff_.data();
+        auto begin = Iterator::begin(const_buff);
+        auto end = Iterator::end(const_buff);
+
+        markers::array_holder_t<Iterator> results;
+        results.elements.reserve(replies_count_);
+        size_t cumulative_consumption = 0;
+
+        do {
+            auto from = begin + cumulative_consumption;
+            auto parse_result = Protocol::parse(from, end);
+            auto *parse_error = boost::get<protocol_error_t>(&parse_result);
+            if (parse_error) {
+                error_code =
+                    Error::make_error_code(bredis_errors::protocol_error);
+                break;
+            } else {
+                auto &positive_result =
+                    boost::get<positive_parse_result_t<Iterator>>(parse_result);
+                results.elements.emplace_back(positive_result.result);
+                cumulative_consumption += positive_result.consumed;
+            }
+        } while (results.elements.size() < replies_count_);
+
+        /* check again, as protocol error might be met */
+        if (!error_code) {
+            if (replies_count_ == 1) {
+                result = positive_parse_result_t<Iterator>{
+                    std::move(results.elements[0]), cumulative_consumption};
+            } else {
+                result = positive_parse_result_t<Iterator>{
+                    std::move(results), cumulative_consumption};
+            }
         }
-        auto &positive_result =
-            boost::get<positive_parse_result_t<Iterator>>(parse_result);
-        results.elements.emplace_back(positive_result.result);
-        cumulative_consumption += positive_result.consumed;
-    } while (results.elements.size() < replies_count_);
-
-    if (replies_count_ == 1) {
-        callback_(error_code, positive_parse_result_t<Iterator>{
-                             std::move(results.elements[0]),
-                             cumulative_consumption});
-    } else {
-        callback_(error_code, positive_parse_result_t<Iterator>{
-                             std::move(results), cumulative_consumption});
     }
+
+    callback_(error_code, std::move(result));
+    done_ = true;
```

Comment:
I don't think you need `done_` at all, Asio will call your `asio_handler_is_continuation` exactly once, when you call `async_read_until`.